passes/opt/opt_expr.cc

```diff
@@ -1251,7 +1251,145 @@ void replace_const_cells(RTLIL::Design *design, RTLIL::Module *module, bool cons
 					}
 			}
 		}
+		if (do_fine && (cell->type.in("$eq", "$ne")))
+		{
+			RTLIL::SigSpec a_sig, b_sig, y_sig;
+			
+			int a_width, b_width;
+			bool a_signed, b_signed;
+			
+			a_sig = assign_map(cell->getPort("\\A"));
+			b_sig = assign_map(cell->getPort("\\B"));
+			y_sig = assign_map(cell->getPort("\\Y"));
+			a_width = cell->parameters["\\A_WIDTH"].as_int();
+			b_width = cell->parameters["\\B_WIDTH"].as_int();
+			a_signed = cell->parameters["\\A_SIGNED"].as_bool();
+			b_signed = cell->parameters["\\B_SIGNED"].as_bool();
+			
+			dict<SigBit, RTLIL::SigSpec> compared_by_const;
+			pool<SigBit> signals_map;
+			SigBit polarity_bit = cell->type == "$eq" ? State::S0 : State::S1;
+
+			if (a_signed || b_signed)
+				goto next_cell;
+
+			for (int i = 0; i < a_width; i++)
+			{
+				if ((a_sig[i] ==  RTLIL::State::S0) || (a_sig[i] ==  RTLIL::State::S1))
+				{
+					if (compared_by_const.count(b_sig[i]))
```

Comment:
What if a_width > b_width? Then b_sig[i] can be out of range. (Same below with reversed roles for a and b.)

Reply:
Resized both to be equal size.